Library/EnvelopeFactory.ts

```diff
@@ -1,13 +1,163 @@
+        const machineName = os.hostname(); // Note: os.hostname() was added in node v0.3.3
+        const instance = (context.tags
+            && context.keys
+            && context.keys.cloudRoleInstance
+            && context.tags[context.keys.cloudRoleInstance]) || machineName;
+
+        var envelope: Contracts.EnvelopeQuickPulse = {
+            Documents: documents.length > 0 ? documents : null,
+            InstrumentationKey: config.instrumentationKey || "",
+            Metrics: metrics.length > 0 ? metrics : null,
+            InvariantVersion: 1, //  1 -> v1 QPS protocol
+            Timestamp: `\/Date(${Date.now()})\/`,
+            Version: context.tags[context.keys.internalSdkVersion],
+            StreamId: StreamId,
+            MachineName: machineName,
+            Instance: instance
+        }
+
+        return envelope;
+    }
+
+    public static createQuickPulseMetric(
+        telemetry: Contracts.MetricTelemetry
+    ): Contracts.MetricQuickPulse {
+        var data: Contracts.MetricQuickPulse;
+        data = {
+            Name: telemetry.name, // TODO: map from MetricTelemetry name to QuickPulse name
+            Value: telemetry.value,
+            Weight: telemetry.count || 1
+        };
+        return data;
+    }
+
+    public static telemetryEnvelopeToQuickPulseDocument(envelope: Contracts.Envelope): Contracts.DocumentQuickPulse {
+        switch (envelope.data.baseType) {
+            case "ExceptionData":
+                return EnvelopeFactory.createQuickPulseExceptionDocument(envelope);
+            case "MessageData":
+                return EnvelopeFactory.createQuickPulseTraceDocument(envelope);
+            case "RemoteDependencyData":
+                return EnvelopeFactory.createQuickPulseDependencyDocument(envelope);
+            case "RequestData":
+                return EnvelopeFactory.createQuickPulseRequestDocument(envelope);
+        }
+        return null;
+    }
+
+    private static createQuickPulseTraceDocument(envelope: Contracts.Envelope): Contracts.MessageDocumentQuickPulse {
+        const document = EnvelopeFactory.createQuickPulseDocument(envelope);
+
+        var traceDocument: Contracts.MessageDocumentQuickPulse = {
+            ...document,
+            Properties: [{key: 'my key', value: 'my value'}],
+            Message: ((envelope.data as any).baseData as Contracts.MessageData).message,
+            SeverityLevel: "Critical"
+        }
+
+        return traceDocument;
+    }
+
+    private static createQuickPulseExceptionDocument(envelope: Contracts.Envelope): Contracts.ExceptionDocumentQuickPulse {
+        const document = EnvelopeFactory.createQuickPulseDocument(envelope);
+
+        let exception = '';
+        ((envelope.data as any).baseData as Contracts.ExceptionData).exceptions[0].parsedStack.forEach(err => {
```

Comment:
Is it possible for `exceptions[0]` to be undefined?

Reply:
Probably only for scenarios like #489, or manual `trackException` calls. Will clean this section up a bit